### PR TITLE
Adding Label Selector Support for Exec

### LINKED ISSUE #8876: kubectl exec by pod label
I often find myself wanting to exec commands on single-container pods, and since pod names are not "stable", I use some wrapper script to avoid constantly updating pod names:

```
kubectl exec -p $(get_current_pod_name) my_script.sh
```

My `get_current_pod_name` returns the name of the first pod matching a label, and that covers the case when I want to execute the script on any one and only one container.

For fetching server logs of different pods/containers, for example, I'm currently giving an argument to `get_current_pod_name` to actually give me the _n-th_ pod from the list of matches, but that's kind ugly... and no multiplexing.

I found this [comment](https://github.com/GoogleCloudPlatform/kubernetes/issues/8448#issuecomment-103235025) by @brendanburns on #8448 which led me to believe that there might be more people who'd like "-l" argument support:

> I would think to do that we'd actually want to use label syntax, e.g.
> 
> ```
> kubectl exec -l stage=production,user=bburns or whatever.
> ```

Would it be reasonable to add label filtering + some way to say "run in any one container" or "run in all containers" (in this case, maybe no support for "-it" flags...)?

### PR DESCRIPTION
**Pull request by joshkurz**
**What this PR does / why we need it**: Add support for label selectors in exec function

**Which issue(s) this PR fixes** :
Fixes #8876

**Special notes for your reviewer**:

Should add some tests cases to prove out the functionality. I made sure nothing that currently exists is broken. Here is an example.

```
kubectl joshua.kurz$ ../../hack/run-in-gopath.sh go run kubectl.go exec -l foo=bar  -- date
Mon Jul 30 07:52:35 UTC 2018
Mon Jul 30 07:52:35 UTC 2018
Mon Jul 30 07:52:35 UTC 2018
Mon Jul 30 07:52:35 UTC 2018
````

Also if you pass in a TTY, it will create one for each pod that is found. If you create a shell in the command it will require you to exit each shell. Thinking we may want to not allow that, but it's also kind of cool that it works. I figure it may be useful and if someone wants to do it they can.

/sig cli

### CODE REVIEW

File: pkg/kubectl/cmd/exec.go [hunk 1]
```diff
@@ -331,3 +342,41 @@ func (p *ExecOptions) Run() error {
 
 	return nil
 }
+
+// Run executes a validated remote execution against a pod or a set of pods.
+func (p *ExecOptions) Run() error {
+	if len(p.Selector) != 0 {
+		items, err := p.PodClient.Pods(p.Namespace).List(metav1.ListOptions{LabelSelector: p.Selector})
+
```
Comment (neolit123) [inline pkg/kubectl/cmd/exec.go]: please, remove this empty line.

File: pkg/kubectl/cmd/exec.go [hunk 2]
```diff
@@ -331,3 +342,41 @@ func (p *ExecOptions) Run() error {
...
+		if err != nil {
+			return err
+		}
+
+		for _, pod := range items.Items {
+			p.Namespace = pod.Namespace
+			p.PodName = pod.Name
+			err = p.Exec()
```
Comment (neolit123) [inline pkg/kubectl/cmd/exec.go]: can be refactored to:
```
if err = p.Exec(); err != nil {
...
```

File: pkg/kubectl/cmd/exec.go [hunk 3]
```diff
@@ -331,3 +342,41 @@ func (p *ExecOptions) Run() error {
...
+
+			if err != nil {
+				return err
+			}
+		}
+
+	} else {
+		pod, err := p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
+
```
Comment (neolit123) [inline pkg/kubectl/cmd/exec.go]: please, remove this empty line.

File: pkg/kubectl/cmd/exec.go [hunk 4]
```diff
@@ -331,3 +342,41 @@ func (p *ExecOptions) Run() error {
...
+		for _, pod := range items.Items {
+			p.Namespace = pod.Namespace
+			p.PodName = pod.Name
+			err = p.Exec()
+
+			if err != nil {
+				return err
+			}
+		}
+
```
Comment (neolit123) [inline pkg/kubectl/cmd/exec.go]: please, remove this empty line.

File: pkg/kubectl/cmd/exec.go [hunk 5]
```diff
@@ -331,3 +342,41 @@ func (p *ExecOptions) Run() error {
...
+		pod, err := p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
+
+		if err != nil {
+			return err
+		}
+
+		p.Namespace = pod.Namespace
+		p.PodName = pod.Name
+		err = p.Exec()
+
```
Comment (neolit123) [inline pkg/kubectl/cmd/exec.go]: can be refactored to:
```
if err = p.Exec(); err != nil {
...
```

File: pkg/kubectl/cmd/exec.go [hunk 6]
```diff
@@ -331,3 +342,41 @@ func (p *ExecOptions) Run() error {
...
+		if err != nil {
+			return err
+		}
+	}
+
```
Comment (neolit123) [inline pkg/kubectl/cmd/exec.go]: please, remove this empty line.

**Review by neolit123 — COMMENTED** [review]
thanks for this PR @joshkurz 
please sign the CLA; more info here: https://github.com/kubernetes/kubernetes/pull/66761#issuecomment-408779175

**Comment by neolit123** [comment]
thanks for signing the CLA. :+1: 
/ok-to-test

**Comment by joshkurz** [comment]
@neolit123 Thanks for the quick review. I assigned @liggitt also, as the bot told me too. :)

If there is anything else you all would like to see let me know. 

Another idea I had was to offer a -p option for parallel. This would allow users to run a command over n pods at the same time. I didn't want to over do it on this first PR, as this new feature is still useful without running commands in parallel.

**Comment by joshkurz** [comment]
looks like `pull-kubernetes-e2e-kops-aws` is failing because it can't find the host.

```
W0805 17:36:52.836] 2018/08/05 17:36:52 process.go:153: Running: kubectl get nodes -ojson
W0805 17:38:24.412] Unable to connect to the server: dial tcp: lookup api.e2e-99757-dba53.test-cncf-aws.k8s.io on 10.63.240.10:53: no such host
W0805 17:38:24.418] 2018/08/05 17:38:24 process.go:155: Step 'kubectl get nodes -ojson' finished in 1m31.639042475s
```

**Comment by liggitt** [comment]
this is clever, but I wonder if it would be better to keep `exec` focused on a single client<->server connection, and invoke it multiple times to accomplish the same result (just like you'd use `ssh` as a building block in a script to run the same command on multiple servers). this seems likely to grow the surface area a lot to accomodate different error handling, short-circuiting, and parallelization options

**Comment by joshkurz** [comment]
Great Points. I do see what you are saying, however I feel like this approach is much cleaner and simpler, and worth the small increase in code responsibility. Having an out of the box solution for this use case I would think is much preferred to the many alternatives.

Anyways, let me know what you think we should do. If nothing, no harm, I learned a lot while doing this, so it's a win win.

**Comment by liggitt** [comment]
I think we'd want to avoid building fan-out and multi-connection handling into a single invocation like this. Let's see what feedback other CLI reviewers have.

@kubernetes/sig-cli-pr-reviews @kubernetes/sig-cli-feature-requests

**Comment by dixudx** [comment]
> I think we'd want to avoid building fan-out and multi-connection handling into a single invocation like this.

Agree. I don't think there is a need to exec into multiple pods at a time. Do we have such a scenario? I don't think so.

Using selectors can help us filter some pods. But such pods may be a single one.

**Comment by smarterclayton** [comment]
I think some fan out scenarios, like logs, are useful for end users day to
day (otherwise we just force someone to reimplement the fan out logic
poorly).  These can be reimplemented elsewhere, but they make trivial
actions like:

1. grab a file out of a set of pods
2. look at the init container pod logs in sequence
3. see all the logs being streamed by a pod right now
4. grep for a specific error in all the pods under a service

fairly annoying.  However, I think if we want to do this we need to put a
bit of thought into it and try to come up with a pattern that is a)
reasonable, b) unlikely to blow up the cluster, and c) amenable to moving
to the server side in the long run.  If we could talk enough about that
pattern to be comfortable with it, it might be something we could
potentially add across a number of commands.

**Comment by joshkurz** [comment]
I agree @smarterclayton. 

One thing I would like to note about this PR. It still needs some more work. Like removing the ability to actually create TTLs across a set of pods. No need for that at all. It also needs to be rebased against master and have some of the new automated github checks accounted for. Making things more reusable is for sure something else that could be worked on.

A specific example, of something else, is for `logs`. `logs -f -l foo-bar` is very beneficial when trying to tail logs across a set of pods and something that is hard to do without logic of this sort integrated into the client. What other commands could benefit from this functionality?

I'd also like to add in the ability to run parallel commands across pods. It's annoying to have to wait for synchronous execution for commands when you have a large amount of pods. 

I'd also like to add the ability to specify a percentage of pods the label selector would match on. So a user could say, give me logs from only 5% of the pods. Or run this command on 5% of the pods.

Curious what your concerns are about blowing up a cluster. I agree if a user runs a command, that is very expensive, they could possibly cause issues. Especially if we add parallel capabilities. However we can't stop people from doing things to themselves by removing power and functionality for users that understand the possibilities of running many commands on cluster at once. Being able to run a command in parallel or synchronous across all pods is a very useful tool in the toolchain for debugging. 

Also could you expand on what you mean by "amenable to moving to the server side in the long run"?

**Comment by kfox1111** [comment]
While I would love to have support for execing across multiple pods, issue #8876 has been open since before Kubernetes 1.0. If it would simplify things, maybe support can be added in a way that only supports a single pod to ease implementation, but could be extended in the future to support execing across multiple pods? Maybe just have it bail with an error message if multiple pods match?

I'd love attach support as well.

**Comment by kfox1111** [comment]
still useful.
/remove-lifecycle rotten

**Comment by joshkurz** [comment]
could you expand on what you mean, when you say attach support?

Just need a little direction in where we want this to go if we want to continue thinking about how to get a solution like this implemented. Willing to pick this back up and get something in place we all agree on. Many ideas have been tossed around on this thread, but nothing really concrete. Want to make sure we are on the same page before continuing with anything.

**Comment by kfox1111** [comment]
When I said attach, I meant to say 'kubectl attach'. Sorry for the confusion. It would be useful to support a label selector to identify the destination pod to attach to, just like its useful to do for 'kubectl exec'.

**Comment by kfox1111** [comment]
still useful.
/remove-lifecycle rotten
/reopen

**Pull request closed without merging**